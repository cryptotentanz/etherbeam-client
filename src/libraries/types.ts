import { BigNumber } from 'ethers'

export const WETH_HASH = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2'
export const USDC_HASH = '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48'

export enum UserType {
  User = 'user',
  Administrator = 'administrator',
}

export enum AddressType {
  Unknown = 'unknown',
  Wallet = 'wallet',
  Token = 'token',
  Contract = 'contract',
  Transaction = 'transaction',
}

export enum TransactionMethodParameterType {
  Unknown = 'unknown',
  UnsignedInteger = 'unsigned_integer',
  Address = 'address',
  Addresses = 'addresses',
}

export enum TransactionStatus {
  Pending = 'pending',
  Mined = 'mined',
  Validated = 'validated',
  Cancelled = 'cancelled',
}

export enum TransactionActionType {
  Unknown = 'unknown',
  Approval = 'approval',
  Transfer = 'transfer',
  Swap = 'swap',
}

export enum TransactionActionDirection {
  Buy = 'buy',
  Sell = 'sell',
}

export interface User {
  name: string
  email: string
  type: UserType
}

export interface TransactionMethodParameter {
  name: string
  type: TransactionMethodParameterType
  rawType: string
  value: TransactionMethodValueType
}

export enum HistoryTime {
  Hour = 'h',
  Day = 'd',
  Week = 'w',
  Month = 'm',
  Year = 'y',
}

export interface TokenPriceHistory {
  time: HistoryTime
  price: BigNumber
  ratio: number
}

export interface Token extends Address {
  name: string
  description?: string
  symbol: string
  decimals: number
  chartPair?: string | null
  price?: BigNumber | null
  priceHistory: TokenPriceHistory[]
  website?: string | null
  whitepaper?: string | null
  github?: string | null
  linkedin?: string | null
  facebook?: string | null
  reddit?: string | null
  twitter?: string | null
  telegram?: string | null
  discord?: string | null
  actions?: TransactionAction[] | null
}

export interface Address {
  hash: string
  url: string
  type: AddressType
  label?: string
}

export interface Transaction {
  blockNumber?: number | null
  dateTime: Date
  address: Address
  status: TransactionStatus
  from?: Address | null
  to?: Address | null
  value?: BigNumber | null
  gas: {
    unitPrice?: BigNumber | null
    limit?: number | null
    used?: number | null
    ratio?: number | null
    price?: BigNumber | null
  }
  action?: TransactionMethod | null
  logs?: TransactionMethod[] | null
}

export interface TransactionMethod {
  contract: Address
  name: string
  parameters: TransactionMethodParameter[]
}

export type TransactionMethodValueType = string | BigNumber | Address | Address[] | null

export interface TransactionAction {
  index: number
  transaction: Transaction
  type: TransactionActionType
  direction?: TransactionActionDirection
  holder: Address
  fromAddress?: Address | Token | null
  fromAmount?: BigNumber | null
  toAddress?: Address | Token | null
  toAmount?: BigNumber | null
}
