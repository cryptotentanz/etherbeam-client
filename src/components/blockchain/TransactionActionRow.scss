@use '../../styles/resources/colors';
@use '../../styles/resources/sizes';

.my-transactionactionrow {
  td {
    padding: sizes.$spacer-xs sizes.$spacer-s !important;
  }

  &-action {
    display: flex;
    align-items: center;
  }

  &-icon {
    margin-right: sizes.$spacer-s;
  }
}
