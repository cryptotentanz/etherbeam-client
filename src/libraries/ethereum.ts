import { Address, AddressType } from './types'

const ETHERSCAN_URL_BASE = 'https://etherscan.io/'

export type EtherscanUrlType = 'block' | 'tx' | 'token' | 'address'

export const getEtherscanUrl = (type: EtherscanUrlType, id: string | number): string => {
  return `${ETHERSCAN_URL_BASE}${type}/${id}`
}

export const createTransactionAddress = (hash: string): Address => {
  return {
    hash,
    type: AddressType.Transaction,
    url: getEtherscanUrl('tx', hash),
  }
}
