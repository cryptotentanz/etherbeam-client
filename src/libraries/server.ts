import compact from 'lodash/compact'
import {
  HttpStatus,
  AuthToken,
  RestAction,
  callApi,
  ApiClient,
  createApiClient,
  setAuthToken,
  subscribeToNewTokens,
  flushAuthToken,
} from '@totentech/api-client'
import { Token, USDC_HASH, User, WETH_HASH } from './types'
import { TokenData, UserData } from './serverTypes'
import { parseTokenData, parseUserData } from './serverParsers'
import { loadAuthToken, removeAuthToken, saveAuthToken } from './storage'

const TIMEOUT = 60 * 1000

type UserUpdateCallback = (user: User | null) => void

export interface ServerData {
  apiClient: ApiClient
  userSubscriptions: UserUpdateCallback[]
}

const onNewAuthToken = (data: ServerData, token: AuthToken): void => {
  saveAuthToken(token)

  if (!token) {
    onUserUpdate(data, null)
  }
}

const initializeAuthToken = (apiClient: ApiClient): void => {
  const token = loadAuthToken()

  if (token) setAuthToken(apiClient, token)
}

export const initializeServerData = (): ServerData => {
  const onUserUpdateCallback = (user: User | null): void => {
    if (!user) removeAuthToken()
  }
  const data = {
    apiClient: createApiClient(process.env.SERVER_API_HOST as string, true, TIMEOUT),
    userSubscriptions: [onUserUpdateCallback],
  }

  initializeAuthToken(data.apiClient)
  subscribeToNewTokens(data.apiClient, (token) => onNewAuthToken(data, token))

  return data
}

export const subscribeToUserUpdates = ({ userSubscriptions }: ServerData, callback: UserUpdateCallback): void => {
  userSubscriptions.push(callback)
}

export const unsubscribeFromUserUpdates = ({ userSubscriptions }: ServerData, callback: UserUpdateCallback): void => {
  const index = userSubscriptions.indexOf(callback)

  if (index < 0) return

  userSubscriptions.splice(index, 1)
}

const onUserUpdate = ({ userSubscriptions }: ServerData, user: User | null): void => {
  userSubscriptions.forEach((callback) => callback(user))
}

export const signIn = async (serverData: ServerData, email: string, password: string): Promise<User | null> => {
  const payload = { email, password }

  const { data, status } = await callApi<UserData>(
    serverData.apiClient,
    RestAction.Post,
    'auth/sign_in',
    [HttpStatus.Ok],
    null,
    payload,
    true
  )

  const user = status != HttpStatus.Unauthorized ? parseUserData(data) : null
  onUserUpdate(serverData, user)

  return user
}

export const validateToken = async (serverData: ServerData): Promise<User | null> => {
  const { data, status } = await callApi<UserData>(
    serverData.apiClient,
    RestAction.Get,
    'auth/validate_token',
    [HttpStatus.Ok],
    null,
    null,
    true
  )

  const user = status != HttpStatus.Unauthorized ? parseUserData(data) : null
  onUserUpdate(serverData, user)

  return user
}

export const signOut = async (data: ServerData): Promise<void> => {
  const { apiClient } = data
  await callApi(apiClient, RestAction.Delete, 'auth/sign_out', [HttpStatus.Ok, HttpStatus.NotFound], null, null, true)

  flushAuthToken(apiClient)
  onUserUpdate(data, null)
}

export const fetchTokens = async (serverData: ServerData): Promise<Token[]> => {
  const { data } = await callApi<TokenData[]>(serverData.apiClient, RestAction.Get, 'contract_tokens')

  const ignoredHashes = [WETH_HASH, USDC_HASH]

  return compact(data.map((token) => (!ignoredHashes.includes(token.address_hash) ? parseTokenData(token) : null)))
}

export const fetchToken = async (
  serverData: ServerData,
  address: string,
  withList?: 'actions'
): Promise<Token | null> => {
  const { status, data } = await callApi<TokenData>(
    serverData.apiClient,
    RestAction.Get,
    `contract_tokens/${address}?list=${withList || ''}`,
    [HttpStatus.Ok, HttpStatus.NotFound]
  )

  if (status != HttpStatus.Ok) return null

  return parseTokenData(data)
}
