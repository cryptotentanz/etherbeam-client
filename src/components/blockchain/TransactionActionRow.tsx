import React, { FC, ReactElement } from 'react'
import { Token, TransactionAction, TransactionActionDirection, TransactionActionType } from 'libraries/types'
import { IconProp } from '@fortawesome/fontawesome-svg-core'
import {
  faArrowAltCircleRight,
  faCheckCircle,
  faMinusCircle,
  faPlusCircle,
  faQuestionCircle,
} from '@fortawesome/free-solid-svg-icons'
import TableRow from '../TableRow'
import Icon from '../Icon'
import Time from '../Time'
import TokenAmount from './TokenAmount'
import Address from './AddressItem'
import EtherscanLink from './EtherscanLink'
import './TransactionActionRow.scss'

interface ActionProps {
  action: TransactionAction
}

const UnknownAction: FC = () => {
  return <span>Unknown action</span>
}

const ApprovalAction: FC = () => {
  return <span>Approved</span>
}

const TransferAction: FC<ActionProps> = ({ action }: ActionProps) => {
  const { fromAmount, fromAddress, toAddress } = action

  return (
    <span>
      <TokenAmount amount={fromAmount} token={fromAddress as Token} />
      <span> to </span>
      {toAddress && <Address address={toAddress} short />}
    </span>
  )
}

const BuyAction: FC<ActionProps> = ({ action }: ActionProps) => {
  const { fromAmount, toAmount, toAddress } = action

  return (
    <span>
      <TokenAmount amount={toAmount} token={toAddress as Token} amountPrintOptions={{ decimals: 0 }} />
      {fromAmount?.isZero() == false && (
        <>
          <span> for </span>
          <TokenAmount amount={fromAmount} amountPrintOptions={{ decimals: 5 }} />
        </>
      )}
    </span>
  )
}

const SellAction: FC<ActionProps> = ({ action }: ActionProps) => {
  const { fromAmount, fromAddress, toAmount } = action
  return (
    <span>
      <TokenAmount amount={fromAmount} token={fromAddress as Token} amountPrintOptions={{ decimals: 0 }} />
      {toAmount?.isZero() == false && (
        <>
          <span> for </span>
          <TokenAmount amount={toAmount} amountPrintOptions={{ decimals: 5 }} />
        </>
      )}
    </span>
  )
}

interface TransactionActionRowProps {
  action: TransactionAction
}

const TransactionActionRow: FC<TransactionActionRowProps> = ({ action }: TransactionActionRowProps) => {
  const {
    type,
    direction,
    transaction: {
      address: { url },
      dateTime,
    },
    holder,
  } = action

  const getProps = (): { color: string; icon: IconProp; Action: ReactElement } => {
    switch (type) {
      case TransactionActionType.Approval:
        return {
          color: 'dark',
          icon: faCheckCircle,
          Action: <ApprovalAction />,
        }
      case TransactionActionType.Transfer:
        return {
          color: 'dark',
          icon: faArrowAltCircleRight,
          Action: <TransferAction action={action} />,
        }
      case TransactionActionType.Swap:
        if (direction == TransactionActionDirection.Buy) {
          return {
            color: 'positive',
            icon: faPlusCircle,
            Action: <BuyAction action={action} />,
          }
        } else {
          return {
            color: 'negative',
            icon: faMinusCircle,
            Action: <SellAction action={action} />,
          }
        }
      default: {
        return {
          color: 'dark',
          icon: faQuestionCircle,
          Action: <UnknownAction />,
        }
      }
    }
  }

  const { color, icon, Action } = getProps()

  return (
    <TableRow className={`my-transactionactionrow my-${color}-bg`}>
      <td>
        <div className="my-transactionactionrow-action">
          <Icon className="my-transactionactionrow-icon my-alt" icon={icon} size="xl" />
          <span>{Action}</span>
        </div>
      </td>
      <td className="my-text-center">
        <Time dateTime={dateTime} />
      </td>
      <td className="my-d-min-xs my-text-center">
        <Address address={holder} short color="light" />
      </td>
      <td className="my-text-center">
        <EtherscanLink url={url} size="l" />
      </td>
    </TableRow>
  )
}

export default TransactionActionRow
