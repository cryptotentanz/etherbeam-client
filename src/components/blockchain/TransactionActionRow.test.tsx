import * as React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'
import { createTokenAction } from 'tests/fixtures/ethereum'
import { TransactionActionDirection, TransactionActionType } from 'libraries/types'
import TransactionActionRow from './TransactionActionRow'

describe('<TransactionActionRow />', () => {
  let component: ShallowWrapper

  describe('Default', () => {
    const action = createTokenAction(TransactionActionType.Unknown)

    beforeAll(() => {
      component = shallow(<TransactionActionRow action={action} />)
    })

    it('should render row', () => expect(component.find('TableRow').hasClass('my-dark-bg')).toBeTruthy())

    it('should render time', () => expect(component.find('Time').prop('dateTime')).toBe(action.transaction.dateTime))
    it('should render address', () => expect(component.find('AddressItem').prop('address')).toBe(action.holder))
    it('should render Etherscan link', () =>
      expect(component.find('EtherscanLink').prop('url')).toBe(action.transaction.address.url))

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Approval', () => {
    const action = createTokenAction(TransactionActionType.Approval)

    beforeAll(() => {
      component = shallow(<TransactionActionRow action={action} />)
    })

    it('should render row', () => expect(component.find('TableRow').hasClass('my-dark-bg')).toBeTruthy())

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Transfer', () => {
    const action = createTokenAction(TransactionActionType.Transfer)

    beforeAll(() => {
      component = shallow(<TransactionActionRow action={action} />)
    })

    it('should render row', () => expect(component.find('TableRow').hasClass('my-dark-bg')).toBeTruthy())

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Buy', () => {
    const action = createTokenAction(TransactionActionType.Swap, TransactionActionDirection.Buy)

    beforeAll(() => {
      component = shallow(<TransactionActionRow action={action} />)
    })

    it('should render row', () => expect(component.find('TableRow').hasClass('my-positive-bg')).toBeTruthy())

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })

  describe('Sell', () => {
    const action = createTokenAction(TransactionActionType.Swap, TransactionActionDirection.Sell)

    beforeAll(() => {
      component = shallow(<TransactionActionRow action={action} />)
    })

    it('should render row', () => expect(component.find('TableRow').hasClass('my-negative-bg')).toBeTruthy())

    it('should match snapshot', () => expect(component.render()).toMatchSnapshot())
  })
})
