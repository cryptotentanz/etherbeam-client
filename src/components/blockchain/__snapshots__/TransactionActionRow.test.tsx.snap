// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TransactionActionRow /> Approval should match snapshot 1`] = `
<tr
  class="my-dark-bg my-transactionactionrow my-dark-bg"
>
  <td>
    <div
      class="my-transactionactionrow-action"
    >
      <span
        class="my-icon my-size-xl my-transactionactionrow-icon my-alt"
      >
        <span
          class=" "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check-circle fa-w-16 "
            data-icon="check-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
      <span>
        <span>
          Approved
        </span>
      </span>
    </div>
  </td>
  <td
    class="my-text-center"
  >
    <span
      class="my-time"
    >
      11:22
    </span>
  </td>
  <td
    class="my-d-min-xs my-text-center"
  >
    <span>
      <a
        class="my-clickable  my-link my-light-fg"
        href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
        rel="noreferrer"
        tabindex="0"
        target="_blank"
      >
        <span
          class="my-badge my-light-bg  my-clickable"
        >
          <span
            class="my-icon  "
          >
            <span
              class=" "
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-wallet fa-w-16 "
                data-icon="wallet"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M461.2 128H80c-8.84 0-16-7.16-16-16s7.16-16 16-16h384c8.84 0 16-7.16 16-16 0-26.51-21.49-48-48-48H64C28.65 32 0 60.65 0 96v320c0 35.35 28.65 64 64 64h397.2c28.02 0 50.8-21.53 50.8-48V176c0-26.47-22.78-48-50.8-48zM416 336c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="my-icon-label "
            >
              My Wallet
            </span>
          </span>
        </span>
      </a>
    </span>
  </td>
  <td
    class="my-text-center"
  >
    <a
      class="my-clickable my-size-l my-ulink my-secondary-fg"
      href="https://etherscan.io/tx/0x0000000000000000000000000000000000000000000000000000000000000001"
      rel="noreferrer"
      tabindex="0"
      target="_blank"
      title="Etherscan"
    >
      <div
        class="my-etherscanlink-img"
      >
        <img
          src="/images/etherscan.svg"
        />
      </div>
    </a>
  </td>
</tr>
`;

exports[`<TransactionActionRow /> Buy should match snapshot 1`] = `
<tr
  class="my-dark-bg my-transactionactionrow my-positive-bg"
>
  <td>
    <div
      class="my-transactionactionrow-action"
    >
      <span
        class="my-icon my-size-xl my-transactionactionrow-icon my-alt"
      >
        <span
          class=" "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-plus-circle fa-w-16 "
            data-icon="plus-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm144 276c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92h-92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
      <span>
        <span>
          <span
            class="my-badge my-light-bg  "
          >
            <span
              class="my-tokenamount-symbol"
            >
              TKN-2
            </span>
            <span
              class="my-tokenamount-amount"
            >
               NaN
            </span>
          </span>
          <span>
             for 
          </span>
          <span
            class="my-badge my-light-bg  "
          >
            <span
              class="my-tokenamount-symbol"
            >
              ETH
            </span>
            <span
              class="my-tokenamount-amount"
            >
               NaN
            </span>
          </span>
        </span>
      </span>
    </div>
  </td>
  <td
    class="my-text-center"
  >
    <span
      class="my-time"
    >
      11:22
    </span>
  </td>
  <td
    class="my-d-min-xs my-text-center"
  >
    <span>
      <a
        class="my-clickable  my-link my-light-fg"
        href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
        rel="noreferrer"
        tabindex="0"
        target="_blank"
      >
        <span
          class="my-badge my-light-bg  my-clickable"
        >
          <span
            class="my-icon  "
          >
            <span
              class=" "
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-wallet fa-w-16 "
                data-icon="wallet"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M461.2 128H80c-8.84 0-16-7.16-16-16s7.16-16 16-16h384c8.84 0 16-7.16 16-16 0-26.51-21.49-48-48-48H64C28.65 32 0 60.65 0 96v320c0 35.35 28.65 64 64 64h397.2c28.02 0 50.8-21.53 50.8-48V176c0-26.47-22.78-48-50.8-48zM416 336c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="my-icon-label "
            >
              My Wallet
            </span>
          </span>
        </span>
      </a>
    </span>
  </td>
  <td
    class="my-text-center"
  >
    <a
      class="my-clickable my-size-l my-ulink my-secondary-fg"
      href="https://etherscan.io/tx/0x0000000000000000000000000000000000000000000000000000000000000001"
      rel="noreferrer"
      tabindex="0"
      target="_blank"
      title="Etherscan"
    >
      <div
        class="my-etherscanlink-img"
      >
        <img
          src="/images/etherscan.svg"
        />
      </div>
    </a>
  </td>
</tr>
`;

exports[`<TransactionActionRow /> Default should match snapshot 1`] = `
<tr
  class="my-dark-bg my-transactionactionrow my-dark-bg"
>
  <td>
    <div
      class="my-transactionactionrow-action"
    >
      <span
        class="my-icon my-size-xl my-transactionactionrow-icon my-alt"
      >
        <span
          class=" "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-question-circle fa-w-16 "
            data-icon="question-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
      <span>
        <span>
          Unknown action
        </span>
      </span>
    </div>
  </td>
  <td
    class="my-text-center"
  >
    <span
      class="my-time"
    >
      11:22
    </span>
  </td>
  <td
    class="my-d-min-xs my-text-center"
  >
    <span>
      <a
        class="my-clickable  my-link my-light-fg"
        href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
        rel="noreferrer"
        tabindex="0"
        target="_blank"
      >
        <span
          class="my-badge my-light-bg  my-clickable"
        >
          <span
            class="my-icon  "
          >
            <span
              class=" "
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-wallet fa-w-16 "
                data-icon="wallet"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M461.2 128H80c-8.84 0-16-7.16-16-16s7.16-16 16-16h384c8.84 0 16-7.16 16-16 0-26.51-21.49-48-48-48H64C28.65 32 0 60.65 0 96v320c0 35.35 28.65 64 64 64h397.2c28.02 0 50.8-21.53 50.8-48V176c0-26.47-22.78-48-50.8-48zM416 336c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="my-icon-label "
            >
              My Wallet
            </span>
          </span>
        </span>
      </a>
    </span>
  </td>
  <td
    class="my-text-center"
  >
    <a
      class="my-clickable my-size-l my-ulink my-secondary-fg"
      href="https://etherscan.io/tx/0x0000000000000000000000000000000000000000000000000000000000000001"
      rel="noreferrer"
      tabindex="0"
      target="_blank"
      title="Etherscan"
    >
      <div
        class="my-etherscanlink-img"
      >
        <img
          src="/images/etherscan.svg"
        />
      </div>
    </a>
  </td>
</tr>
`;

exports[`<TransactionActionRow /> Sell should match snapshot 1`] = `
<tr
  class="my-dark-bg my-transactionactionrow my-negative-bg"
>
  <td>
    <div
      class="my-transactionactionrow-action"
    >
      <span
        class="my-icon my-size-xl my-transactionactionrow-icon my-alt"
      >
        <span
          class=" "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-minus-circle fa-w-16 "
            data-icon="minus-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
      <span>
        <span>
          <span
            class="my-badge my-light-bg  "
          >
            <span
              class="my-tokenamount-symbol"
            >
              TKN-1
            </span>
            <span
              class="my-tokenamount-amount"
            >
               NaN
            </span>
          </span>
          <span>
             for 
          </span>
          <span
            class="my-badge my-light-bg  "
          >
            <span
              class="my-tokenamount-symbol"
            >
              ETH
            </span>
            <span
              class="my-tokenamount-amount"
            >
               NaN
            </span>
          </span>
        </span>
      </span>
    </div>
  </td>
  <td
    class="my-text-center"
  >
    <span
      class="my-time"
    >
      11:22
    </span>
  </td>
  <td
    class="my-d-min-xs my-text-center"
  >
    <span>
      <a
        class="my-clickable  my-link my-light-fg"
        href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
        rel="noreferrer"
        tabindex="0"
        target="_blank"
      >
        <span
          class="my-badge my-light-bg  my-clickable"
        >
          <span
            class="my-icon  "
          >
            <span
              class=" "
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-wallet fa-w-16 "
                data-icon="wallet"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M461.2 128H80c-8.84 0-16-7.16-16-16s7.16-16 16-16h384c8.84 0 16-7.16 16-16 0-26.51-21.49-48-48-48H64C28.65 32 0 60.65 0 96v320c0 35.35 28.65 64 64 64h397.2c28.02 0 50.8-21.53 50.8-48V176c0-26.47-22.78-48-50.8-48zM416 336c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="my-icon-label "
            >
              My Wallet
            </span>
          </span>
        </span>
      </a>
    </span>
  </td>
  <td
    class="my-text-center"
  >
    <a
      class="my-clickable my-size-l my-ulink my-secondary-fg"
      href="https://etherscan.io/tx/0x0000000000000000000000000000000000000000000000000000000000000001"
      rel="noreferrer"
      tabindex="0"
      target="_blank"
      title="Etherscan"
    >
      <div
        class="my-etherscanlink-img"
      >
        <img
          src="/images/etherscan.svg"
        />
      </div>
    </a>
  </td>
</tr>
`;

exports[`<TransactionActionRow /> Transfer should match snapshot 1`] = `
<tr
  class="my-dark-bg my-transactionactionrow my-dark-bg"
>
  <td>
    <div
      class="my-transactionactionrow-action"
    >
      <span
        class="my-icon my-size-xl my-transactionactionrow-icon my-alt"
      >
        <span
          class=" "
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-arrow-alt-circle-right fa-w-16 "
            data-icon="arrow-alt-circle-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 8c137 0 248 111 248 248S393 504 256 504 8 393 8 256 119 8 256 8zM140 300h116v70.9c0 10.7 13 16.1 20.5 8.5l114.3-114.9c4.7-4.7 4.7-12.2 0-16.9l-114.3-115c-7.6-7.6-20.5-2.2-20.5 8.5V212H140c-6.6 0-12 5.4-12 12v64c0 6.6 5.4 12 12 12z"
              fill="currentColor"
            />
          </svg>
        </span>
      </span>
      <span>
        <span>
          <span
            class="my-badge my-light-bg  "
          >
            <span
              class="my-tokenamount-symbol"
            >
              TKN-1
            </span>
            <span
              class="my-tokenamount-amount"
            >
               NaN
            </span>
          </span>
          <span>
             to 
          </span>
          <span>
            <a
              class="my-clickable  my-link my-secondary-fg"
              href="https://etherscan.io/address/0x0000000000000000000000000000000000000222"
              rel="noreferrer"
              tabindex="0"
              target="_blank"
            >
              <span
                class="my-badge my-light-bg  my-clickable"
              >
                <span
                  class="my-icon  "
                >
                  <span
                    class=" "
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-coins fa-w-16 "
                      data-icon="coins"
                      data-prefix="fas"
                      focusable="false"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M0 405.3V448c0 35.3 86 64 192 64s192-28.7 192-64v-42.7C342.7 434.4 267.2 448 192 448S41.3 434.4 0 405.3zM320 128c106 0 192-28.7 192-64S426 0 320 0 128 28.7 128 64s86 64 192 64zM0 300.4V352c0 35.3 86 64 192 64s192-28.7 192-64v-51.6c-41.3 34-116.9 51.6-192 51.6S41.3 334.4 0 300.4zm416 11c57.3-11.1 96-31.7 96-55.4v-42.7c-23.2 16.4-57.3 27.6-96 34.5v63.6zM192 160C86 160 0 195.8 0 240s86 80 192 80 192-35.8 192-80-86-80-192-80zm219.3 56.3c60-10.8 100.7-32 100.7-56.3v-42.7c-35.5 25.1-96.5 38.6-160.7 41.8 29.5 14.3 51.2 33.5 60 57.2z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                  <span
                    class="my-icon-label "
                  >
                    Token 2 (TKN-2)
                  </span>
                </span>
              </span>
            </a>
          </span>
        </span>
      </span>
    </div>
  </td>
  <td
    class="my-text-center"
  >
    <span
      class="my-time"
    >
      11:22
    </span>
  </td>
  <td
    class="my-d-min-xs my-text-center"
  >
    <span>
      <a
        class="my-clickable  my-link my-light-fg"
        href="https://etherscan.io/address/0x0000000000000000000000000000000000000111"
        rel="noreferrer"
        tabindex="0"
        target="_blank"
      >
        <span
          class="my-badge my-light-bg  my-clickable"
        >
          <span
            class="my-icon  "
          >
            <span
              class=" "
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-wallet fa-w-16 "
                data-icon="wallet"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M461.2 128H80c-8.84 0-16-7.16-16-16s7.16-16 16-16h384c8.84 0 16-7.16 16-16 0-26.51-21.49-48-48-48H64C28.65 32 0 60.65 0 96v320c0 35.35 28.65 64 64 64h397.2c28.02 0 50.8-21.53 50.8-48V176c0-26.47-22.78-48-50.8-48zM416 336c-17.67 0-32-14.33-32-32s14.33-32 32-32 32 14.33 32 32-14.33 32-32 32z"
                  fill="currentColor"
                />
              </svg>
            </span>
            <span
              class="my-icon-label "
            >
              My Wallet
            </span>
          </span>
        </span>
      </a>
    </span>
  </td>
  <td
    class="my-text-center"
  >
    <a
      class="my-clickable my-size-l my-ulink my-secondary-fg"
      href="https://etherscan.io/tx/0x0000000000000000000000000000000000000000000000000000000000000001"
      rel="noreferrer"
      tabindex="0"
      target="_blank"
      title="Etherscan"
    >
      <div
        class="my-etherscanlink-img"
      >
        <img
          src="/images/etherscan.svg"
        />
      </div>
    </a>
  </td>
</tr>
`;
